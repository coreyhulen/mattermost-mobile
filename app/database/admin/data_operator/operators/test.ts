// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Q} from '@nozbe/watermelondb';

import {MM_TABLES} from '@constants/database';
import App from '@typings/database/app';

import DatabaseManager, {DatabaseType} from '../../database_manager';
import DataOperator from '../index';
import {IsolatedEntities, OperationType} from '../types';
import {
    operateAppRecord,
    operateCustomEmojiRecord,
    operateGlobalRecord,
    operateRoleRecord,
    operateServersRecord,
    operateSystemRecord,
    operateTermsOfServiceRecord,
} from './index';

jest.mock('../../database_manager');

const {APP} = MM_TABLES.DEFAULT;

describe('*** DataOperator: Operators tests ***', () => {
    const createConnection = async (setActive = false) => {
        const dbName = 'server_schema_connection';
        const serverUrl = 'https://appv2.mattermost.com';
        const database = await DatabaseManager.createDatabaseConnection({
            shouldAddToDefaultDatabase: true,
            databaseConnection: {
                actionsEnabled: true,
                dbName,
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
        });

        if (setActive) {
            await DatabaseManager.setActiveServerDatabase({
                displayName: dbName,
                serverUrl,
            });
        }

        return database;
    };

    it('=> operateAppRecord: should return an array of type App ', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateAppRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {
                buildNumber: 'build-7',
                createdAt: 1,
                id: 'id-18',
                versionNumber: 'v-1',
            },
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('App');
    });

    it('=> operateGlobalRecord: should return an array of type Global', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateGlobalRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'g-1', name: 'g-n1', value: 'g-v1'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Global');
    });

    it('=> operateServersRecord: should return an array of type Servers', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateServersRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {
                dbPath: 'mm-server',
                displayName: 's-displayName',
                id: 's-1',
                mentionCount: 1,
                unreadCount: 0,
                url: 'https://community.mattermost.com',
            },
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Servers');
    });

    it('=> operateCustomEmojiRecord: should return an array of type CustomEmoji', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateCustomEmojiRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'emo-1', name: 'emoji'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('CustomEmoji');
    });

    it('=> operateRoleRecord: should return an array of type Role ', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateRoleRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'role-1', name: 'role-name-1', permissions: []},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Role');
    });

    it('=> operateSystemRecord: should return an array of type System', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateSystemRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'system-1', name: 'system-name-1', value: 'system'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('System');
    });

    it('=> operateTermsOfServiceRecord: should return an array of type TermsOfService ', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateTermsOfServiceRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'system-1', acceptedAt: 1},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch(
            'TermsOfService',
        );
    });

    it('=> should create a record in the App table in the default database', async () => {
        expect.assertions(2);

        // Creates a record in the App table
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-1',
                    createdAt: 1,
                    id: 'id-1',
                    versionNumber: 'version-1',
                },
            ],
        });

        // Do a query and find out if the value has been registered in the App table of the default database
        const connection = await DatabaseManager.getDefaultDatabase();
        expect(connection).toBeTruthy();

        const records = (await connection!.collections.
            get(APP).
            query(Q.where('id', 'id-1')).
            fetch()) as App[];

        // We should expect to have a record returned as dictated by our query
        expect(records.length).toBe(1);
    });

    it('=> should create several records in the App table in the default database', async () => {
        expect.assertions(2);

        // Creates a record in the App table
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
                {
                    buildNumber: 'build-12',
                    createdAt: 1,
                    id: 'id-12',
                    versionNumber: 'version-12',
                },
                {
                    buildNumber: 'build-13',
                    createdAt: 1,
                    id: 'id-13',
                    versionNumber: 'version-13',
                },
            ],
        });

        // Do a query and find out if the value has been registered in the App table of the default database
        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11', 'id-12', 'id-13']))).
            fetch()) as App[];

        // We should expect to have 4 records created
        expect(records.length).toBe(4);
    });

    it('=> should update a record in the App table in the default database', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // Update record having id 'id-1'
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-13-13',
                    createdAt: 1,
                    id: 'id-1',
                    versionNumber: 'version-1',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', 'id-1')).
            fetch()) as App[];
        expect(records.length).toBeGreaterThan(0);

        // Verify if the buildNumber for this record has been updated
        expect(records[0].buildNumber).toMatch('build-13-13');
    });

    it('=> should update several records in the App table in the default database', async () => {
        expect.assertions(4);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // Update records having id 'id-10' and 'id-11'
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11y',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11']))).
            fetch()) as App[];
        expect(records.length).toBe(2);

        // Verify if the buildNumber for those two record has been updated
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11y');
    });

    it('=> [EDGE CASE] should UPDATE instead of CREATE record for existing id', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // id-10 and id-11 exist but yet the optType is CREATE.  The operator should then prepareUpdate the records instead of prepareCreate
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11x',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11']))).
            fetch()) as App[];

        // Verify if the buildNumber for those two record has been updated
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11x');
    });

    it('=> [EDGE CASE] should CREATE instead of UPDATE record for non-existing id', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // id-15 and id-16 do not exist but yet the optType is UPDATE.  The operator should then prepareCreate the records instead of prepareUpdate
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-15',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11x',
                    createdAt: 1,
                    id: 'id-16',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-15', 'id-16']))).
            fetch()) as App[];

        // Verify if the buildNumber for those two record has been created
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11x');
    });

    // TODO :  new operators
});
