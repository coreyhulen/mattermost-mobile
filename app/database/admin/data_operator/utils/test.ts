// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {mockedPosts} from './mock';
import {createPostsChain, sanitizePosts} from './utils';

describe('DataOperator: Utils tests', () => {
    it('sanitizePosts: should filter between ordered and unordered posts', () => {
        const {orderedPosts, unOrderedPosts} = sanitizePosts({posts: Object.values(mockedPosts.posts), orders: mockedPosts.order});
        expect(orderedPosts.length).toBe(4);
        expect(unOrderedPosts.length).toBe(2);
    });

    it('createPostsChain: should link posts amongst each other based on order array', () => {
        const previousPostId = 'prev_xxyuoxmehne';
        const chainedOfPosts = createPostsChain({orders: mockedPosts.order, rawPosts: Object.values(mockedPosts.posts), previousPostId});

        // eslint-disable-next-line max-nested-callbacks
        const post1 = chainedOfPosts.find((post) => {
            return post.id === '8swgtrrdiff89jnsiwiip3y1eoe';
        });
        expect(post1).toBeTruthy();
        expect(post1!.prev_post_id).toBe(previousPostId);

        // eslint-disable-next-line max-nested-callbacks
        const post2 = chainedOfPosts.find((post) => {
            return post.id === '8fcnk3p1jt8mmkaprgajoxz115a';
        });
        expect(post2).toBeTruthy();
        expect(post2!.prev_post_id).toBe('8swgtrrdiff89jnsiwiip3y1eoe');

        // eslint-disable-next-line max-nested-callbacks
        const post3 = chainedOfPosts.find((post) => {
            return post.id === '3y3w3a6gkbg73bnj3xund9o5ic';
        });
        expect(post3).toBeTruthy();
        expect(post3!.prev_post_id).toBe('8fcnk3p1jt8mmkaprgajoxz115a');

        // eslint-disable-next-line max-nested-callbacks
        const post4 = chainedOfPosts.find((post) => {
            return post.id === '4btbnmticjgw7ewd3qopmpiwqw';
        });
        expect(post4).toBeTruthy();
        expect(post4!.prev_post_id).toBe('3y3w3a6gkbg73bnj3xund9o5ic');
    });
});
