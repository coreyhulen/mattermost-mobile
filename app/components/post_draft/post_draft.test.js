// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {Alert} from 'react-native';
import assert from 'assert';
import {shallowWithIntl} from 'test/intl-test-helper';

import Preferences from '@mm-redux/constants/preferences';
import PostDraft from './post_draft';

jest.mock('react-native-image-picker', () => ({
    launchCamera: jest.fn(),
}));

describe('PostDraft', () => {
    const baseProps = {
        addReactionToLatestPost: jest.fn(),
        createPost: jest.fn(),
        executeCommand: jest.fn(),
        handleCommentDraftChanged: jest.fn(),
        handlePostDraftChanged: jest.fn(),
        handleClearFiles: jest.fn(),
        handleClearFailedFiles: jest.fn(),
        handleRemoveLastFile: jest.fn(),
        initUploadFiles: jest.fn(),
        userTyping: jest.fn(),
        handleCommentDraftSelectionChanged: jest.fn(),
        setStatus: jest.fn(),
        selectPenultimateChannel: jest.fn(),
        getChannelTimezones: jest.fn(),
        getChannelMemberCountsByGroup: jest.fn(),
        canUploadFiles: true,
        channelId: 'channel-id',
        channelDisplayName: 'Test Channel',
        channelTeamId: 'channel-team-id',
        channelIsReadOnly: false,
        currentUserId: 'current-user-id',
        deactivatedChannel: false,
        files: [],
        maxFileSize: 1024,
        maxMessageLength: 4000,
        rootId: '',
        theme: Preferences.THEMES.default,
        uploadFileRequestStatus: 'NOT_STARTED',
        value: '',
        userIsOutOfOffice: false,
        channelIsArchived: false,
        onCloseChannel: jest.fn(),
        cursorPositionEvent: '',
        valueEvent: '',
        isLandscape: false,
        screenId: 'NavigationScreen1',
        canPost: true,
        currentChannelMembersCount: 50,
        enableConfirmNotificationsToChannel: true,
        useChannelMentions: true,
        useGroupMentions: true,
        groupsWithAllowReference: new Map([
            ['@developers', {
                id: 'developers',
                name: 'developers',
            }],
            ['@qa', {
                id: 'qa',
                name: 'qa',
            }],
        ]),
        channelMemberCountsByGroup: {
            developers: {
                channel_member_count: 10,
                channel_member_timezones_count: 0,
            },
            qa: {
                channel_member_count: 3,
                channel_member_timezones_count: 0,
            },
        },
        membersCount: 10,
    };
    const ref = React.createRef();

    test('should send an alert when sending a message with a channel mention', () => {
        const wrapper = shallowWithIntl(
            <PostDraft
                {...baseProps}
                ref={ref}
            />,
        );
        const message = '@all';
        const instance = wrapper.instance();
        expect(instance.input).toEqual({current: null});
        instance.input = {
            current: {
                getValue: () => message,
                setValue: jest.fn(),
                changeDraft: jest.fn(),
            },
        };

        instance.sendMessage();
        expect(Alert.alert).toBeCalled();
        expect(Alert.alert).toHaveBeenCalledWith('Confirm sending notifications to entire channel', expect.anything(), expect.anything());
    });

    test('should send an alert when sending a message with a group mention with group with count more than NOTIFY_ALL', () => {
        const wrapper = shallowWithIntl(
            <PostDraft
                {...baseProps}
                ref={ref}
            />,
        );
        const message = '@developers';
        const instance = wrapper.instance();
        expect(instance.input).toEqual({current: null});
        instance.input = {
            current: {
                getValue: () => message,
                setValue: jest.fn(),
                changeDraft: jest.fn(),
            },
        };
        instance.sendMessage();
        expect(Alert.alert).toBeCalled();
    });

    test('should not send an alert when sending a message with a group mention with group with count less than NOTIFY_ALL', () => {
        const wrapper = shallowWithIntl(
            <PostDraft
                {...baseProps}
                ref={ref}
            />,
        );
        const message = '@qa';
        const instance = wrapper.instance();
        expect(instance.input).toEqual({current: null});
        instance.input = {
            current: {
                getValue: () => message,
                setValue: jest.fn(),
                changeDraft: jest.fn(),
            },
        };

        instance.sendMessage();
        expect(Alert.alert).not.toBeCalled();
    });

    test('should not send an alert when sending a message with a channel mention when the user does not have channel mentions permission', () => {
        const wrapper = shallowWithIntl(
            <PostDraft
                {...baseProps}
                useChannelMentions={false}
                ref={ref}
            />,
        );
        const message = '@all';
        const instance = wrapper.instance();
        expect(instance.input).toEqual({current: null});
        instance.input = {
            current: {
                getValue: () => message,
                setValue: jest.fn(),
                changeDraft: jest.fn(),
            },
        };

        instance.sendMessage();
        expect(Alert.alert).not.toHaveBeenCalled();
    });

    test('should not send an alert when sending a message with a channel mention when the user does not have group mentions permission', () => {
        const wrapper = shallowWithIntl(
            <PostDraft
                {...baseProps}
                useGroupMentions={false}
                ref={ref}
            />,
        );
        const message = '@developer';
        const instance = wrapper.instance();
        expect(instance.input).toEqual({current: null});
        instance.input = {
            current: {
                getValue: () => message,
                setValue: jest.fn(),
                changeDraft: jest.fn(),
            },
        };

        instance.sendMessage();
        expect(Alert.alert).not.toHaveBeenCalled();
    });

    test('should return correct @all (same for @channel)', () => {
        for (const data of [
            {
                text: '',
                result: false,
            },
            {
                text: 'all',
                result: false,
            },
            {
                text: '@allison',
                result: false,
            },
            {
                text: '@ALLISON',
                result: false,
            },
            {
                text: '@all123',
                result: false,
            },
            {
                text: '123@all',
                result: false,
            },
            {
                text: 'hey@all',
                result: false,
            },
            {
                text: 'hey@all.com',
                result: false,
            },
            {
                text: '@all',
                result: true,
            },
            {
                text: '@ALL',
                result: true,
            },
            {
                text: '@all hey',
                result: true,
            },
            {
                text: 'hey @all',
                result: true,
            },
            {
                text: 'HEY @ALL',
                result: true,
            },
            {
                text: 'hey @all!',
                result: true,
            },
            {
                text: 'hey @all:+1:',
                result: true,
            },
            {
                text: 'hey @ALL:+1:',
                result: true,
            },
            {
                text: '`@all`',
                result: false,
            },
            {
                text: '@someone `@all`',
                result: false,
            },
            {
                text: '``@all``',
                result: false,
            },
            {
                text: '```@all```',
                result: false,
            },
            {
                text: '```\n@all\n```',
                result: false,
            },
            {
                text: '```````\n@all\n```````',
                result: false,
            },
            {
                text: '```code\n@all\n```',
                result: false,
            },
            {
                text: '~~~@all~~~',
                result: true,
            },
            {
                text: '~~~\n@all\n~~~',
                result: false,
            },
            {
                text: ' /not_cmd @all',
                result: true,
            },
            {
                text: '@channel',
                result: true,
            },
            {
                text: '@channel.',
                result: true,
            },
            {
                text: '@channel/test',
                result: true,
            },
            {
                text: 'test/@channel',
                result: true,
            },
            {
                text: '@all/@channel',
                result: true,
            },
            {
                text: '@cha*nnel*',
                result: false,
            },
            {
                text: '@cha**nnel**',
                result: false,
            },
            {
                text: '*@cha*nnel',
                result: false,
            },
            {
                text: '[@chan](https://google.com)nel',
                result: false,
            },
            {
                text: '@cha![](https://myimage)nnel',
                result: false,
            },
        ]) {
            const wrapper = shallowWithIntl(
                <PostDraft {...baseProps}/>,
            );
            const containsAtChannel = wrapper.instance().textContainsAtAllAtChannel(data.text);
            assert.equal(containsAtChannel, data.result, data.text);
        }
    });
});
