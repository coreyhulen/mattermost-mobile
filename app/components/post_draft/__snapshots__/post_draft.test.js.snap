// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDraft Should render the Archived for channelIsArchived 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderTopColor": "rgba(61,60,64,0.2)",
      "borderTopWidth": 1,
      "paddingBottom": 10,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 10,
    }
  }
>
  <Text>
    <Text
      style={
        Array [
          Object {
            "color": "#3d3c40",
            "textAlign": "center",
          },
          Array [
            undefined,
          ],
        ]
      }
    >
      You are viewing an 
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#3d3c40",
            "textAlign": "center",
          },
          Array [
            undefined,
            Object {
              "fontWeight": "bold",
            },
          ],
        ]
      }
    >
      archived channel
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#3d3c40",
            "textAlign": "center",
          },
          Array [
            undefined,
          ],
        ]
      }
    >
      . New messages cannot be posted.
    </Text>
  </Text>
  <View
    accessibilityRole="button"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#166de0",
        "borderRadius": 4,
        "height": 40,
        "marginTop": 10,
        "opacity": 1,
        "paddingBottom": 5,
        "paddingTop": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontWeight": "bold",
          "marginTop": 7,
        }
      }
    >
      Close Channel
    </Text>
  </View>
</View>
`;

exports[`PostDraft Should render the Archived for deactivatedChannel 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderTopColor": "rgba(61,60,64,0.2)",
      "borderTopWidth": 1,
      "paddingBottom": 10,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 10,
    }
  }
>
  <Text>
    <Text
      style={
        Array [
          Object {
            "color": "#3d3c40",
            "textAlign": "center",
          },
          Array [
            undefined,
          ],
        ]
      }
    >
      You are viewing an 
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#3d3c40",
            "textAlign": "center",
          },
          Array [
            undefined,
            Object {
              "fontWeight": "bold",
            },
          ],
        ]
      }
    >
      archived channel
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#3d3c40",
            "textAlign": "center",
          },
          Array [
            undefined,
          ],
        ]
      }
    >
      . New messages cannot be posted.
    </Text>
  </Text>
  <View
    accessibilityRole="button"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#166de0",
        "borderRadius": 4,
        "height": 40,
        "marginTop": 10,
        "opacity": 1,
        "paddingBottom": 5,
        "paddingTop": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontWeight": "bold",
          "marginTop": 7,
        }
      }
    >
      Close Channel
    </Text>
  </View>
</View>
`;

exports[`PostDraft Should render the DraftInput 1`] = `
<KeyboardTrackingView>
  <View
    style={
      Object {
        "bottom": 0,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "flex": 1,
          "marginLeft": 0,
          "marginRight": 0,
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "backgroundColor": "transparent",
            "color": "#3d3c40",
            "fontSize": 11,
            "marginBottom": 2,
            "paddingLeft": 10,
            "paddingTop": 3,
            "position": "absolute",
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#ffffff",
            "height": 0,
          }
        }
      />
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "backgroundColor": "#ffffff",
          "borderTopColor": "rgba(61,60,64,0.2)",
          "borderTopWidth": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 2,
        },
        null,
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "alignItems": "stretch",
          "paddingTop": 7,
        }
      }
      disableScrollViewPanResponder={true}
      keyboardShouldPersistTaps="always"
      overScrollMode="never"
      pinchGestureEnabled={false}
      scrollEnabled={false}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <View>
        <TextInput
          allowFontScaling={true}
          blurOnSubmit={false}
          disableFullscreenUI={true}
          keyboardAppearance="light"
          keyboardType="default"
          multiline={true}
          onChangeText={[Function]}
          onEndEditing={[Function]}
          onPaste={[Function]}
          onSelectionChange={[Function]}
          placeholder="Write to "
          placeholderTextColor="rgba(61,60,64,0.5)"
          rejectResponderTermination={true}
          style={
            Object {
              "color": "#3d3c40",
              "fontSize": 15,
              "lineHeight": 20,
              "maxHeight": 150,
              "minHeight": 30,
              "paddingBottom": 6,
              "paddingHorizontal": 12,
              "paddingTop": 6,
            }
          }
          testID="post_input"
          underlineColorAndroid="transparent"
        />
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "column",
            }
          }
        >
          <View
            forwardedRef={[Function]}
            isInteraction={true}
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "height": 0,
                "paddingBottom": 0,
              }
            }
          >
            <RCTScrollView
              contentContainerStyle={
                Object {
                  "alignItems": "flex-end",
                  "paddingRight": 12,
                }
              }
              horizontal={true}
              keyboardShouldPersistTaps="handled"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View />
            </RCTScrollView>
          </View>
          <View
            forwardedRef={[Function]}
            isInteraction={true}
            style={
              Object {
                "height": 0,
              }
            }
          >
            <View
              forwardedRef={[Function]}
              isInteraction={true}
              style={
                Object {
                  "flex": 1,
                  "marginHorizontal": 12,
                  "marginTop": 4,
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingBottom": 1,
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "height": 44,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "opacity": 1,
                  "padding": 10,
                }
              }
            >
              <Icon
                color="rgba(61,60,64,0.64)"
                name="at"
                size={24}
              />
            </View>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "opacity": 1,
                  "padding": 10,
                }
              }
            >
              <Icon
                color="rgba(61,60,64,0.64)"
                name="slash-forward-box-outline"
                size={24}
              />
            </View>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "opacity": 1,
                  "padding": 10,
                }
              }
            >
              <Icon
                color="rgba(61,60,64,0.64)"
                name="file-document-outline"
                size={24}
              />
            </View>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "opacity": 1,
                  "padding": 10,
                }
              }
            >
              <Icon
                color="rgba(61,60,64,0.64)"
                name="image-outline"
                size={24}
              />
            </View>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "opacity": 1,
                  "padding": 10,
                }
              }
            >
              <Icon
                color="rgba(61,60,64,0.64)"
                name="camera-outline"
                size={24}
              />
            </View>
          </View>
          <View
            style={
              Object {
                "justifyContent": "flex-end",
                "paddingRight": 8,
              }
            }
            testID="disabled_send_button"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#166de0",
                    "borderRadius": 4,
                    "height": 32,
                    "justifyContent": "center",
                    "width": 80,
                  },
                  Object {
                    "backgroundColor": "rgba(22,109,224,0.3)",
                  },
                ]
              }
            >
              <Icon
                color="rgba(255,255,255,0.5)"
                name="send"
                size={24}
              />
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</KeyboardTrackingView>
`;

exports[`PostDraft Should render the ReadOnly for canPost 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "rgba(61,60,64,0.04)",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderTopColor": "rgba(61,60,64,0.2)",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "height": 50,
        "paddingHorizontal": 12,
      }
    }
  >
    <Icon
      color="#3d3c40"
      name="glasses"
      style={
        Object {
          "fontSize": 20,
          "lineHeight": 22,
          "opacity": 0.56,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#3d3c40",
          "fontSize": 15,
          "lineHeight": 20,
          "marginLeft": 9,
          "opacity": 0.56,
        }
      }
    >
      This channel is read-only.
    </Text>
  </View>
</RCTSafeAreaView>
`;

exports[`PostDraft Should render the ReadOnly for channelIsReadOnly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "rgba(61,60,64,0.04)",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderTopColor": "rgba(61,60,64,0.2)",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "height": 50,
        "paddingHorizontal": 12,
      }
    }
  >
    <Icon
      color="#3d3c40"
      name="glasses"
      style={
        Object {
          "fontSize": 20,
          "lineHeight": 22,
          "opacity": 0.56,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#3d3c40",
          "fontSize": 15,
          "lineHeight": 20,
          "marginLeft": 9,
          "opacity": 0.56,
        }
      }
    >
      This channel is read-only.
    </Text>
  </View>
</RCTSafeAreaView>
`;
